test/Microsoft.AspNetCore.Razor.Language.Test/CodeGeneration/TagHelperHtmlAttributeRuntimeBasicWriterTest.cs

```diff
@@ -0,0 +1,135 @@
+
+            var content = "<input checked=\"hello-world @false\" />";
+            var sourceDocument = TestRazorSourceDocument.Create(content);
+            var codeDocument = RazorCodeDocument.Create(sourceDocument);
+            var irDocument = Lower(codeDocument);
+            var node = irDocument.Children.OfType<HtmlAttributeIRNode>().Single().Children[0] as HtmlAttributeValueIRNode;
+
+            // Act
+            writer.WriteHtmlAttributeValue(context, node);
+
+            // Assert
+            var csharp = context.Writer.Builder.ToString();
+            Assert.Equal(
+@"AddHtmlAttributeValue("""", 16, ""hello-world"", 16, 11, true);
+",
+                csharp,
+                ignoreLineEndingDifferences: true);
+        }
+
+        [Fact]
+        public void WriteCSharpAttributeValue_RendersCorrectly()
+        {
+            var writer = new TagHelperHtmlAttributeRuntimeBasicWriter();
+            var context = GetCSharpRenderingContext(writer);
+
+            var content = "<input checked=\"hello-world @false\" />";
+            var sourceDocument = TestRazorSourceDocument.Create(content);
+            var codeDocument = RazorCodeDocument.Create(sourceDocument);
+            var irDocument = Lower(codeDocument);
+            var node = irDocument.Children.OfType<HtmlAttributeIRNode>().Single().Children[1] as CSharpAttributeValueIRNode;
+
+            // Act
+            writer.WriteCSharpAttributeValue(context, node);
+
+            // Assert
+            var csharp = context.Writer.Builder.ToString();
+            Assert.Equal(
+@"#line 1 ""test.cshtml""
+AddHtmlAttributeValue("" "", 27, false, 28, 6, false);
+
+#line default
+#line hidden
+",
+                csharp,
+                ignoreLineEndingDifferences: true);
+        }
+
+        [Fact]
+        public void WriteCSharpAttributeValue_NonExpression_BuffersResult()
+        {
+            var writer = new TagHelperHtmlAttributeRuntimeBasicWriter();
+            var context = GetCSharpRenderingContext(writer);
+
+            var content = "<input checked=\"hello-world @if(@true){ }\" />";
+            var sourceDocument = TestRazorSourceDocument.Create(content);
+            var codeDocument = RazorCodeDocument.Create(sourceDocument);
+            var irDocument = Lower(codeDocument);
+            var node = irDocument.Children.OfType<HtmlAttributeIRNode>().Single().Children[1] as CSharpAttributeValueIRNode;
+
+            // Act
+            writer.WriteCSharpAttributeValue(context, node);
+
+            // Assert
+            var csharp = context.Writer.Builder.ToString();
+            Assert.Equal(
+@"AddHtmlAttributeValue("" "", 27, new Microsoft.AspNetCore.Mvc.Razor.HelperResult(async(__razor_attribute_value_writer) => {
+    Render Children
+}
+), 28, 13, false);
+",
+                csharp,
+                ignoreLineEndingDifferences: true);
+        }
+
+        private static CSharpRenderingContext GetCSharpRenderingContext(BasicWriter writer)
```

Comment:
Any possibility to make a test base class to hold all these infrastructure bits? It looks like they're duplicated quite a bit